cmd/plume/prerelease.go

```diff
@@ -264,3 +280,253 @@ func azurePreRelease(ctx context.Context, client *http.Client, src *storage.Buck
 
 	return nil
 }
+
+func awsUploadToCloud(spec *channelSpec, cloud *awsCloudSpec, imageName, imageDescription, imagePath string) (map[string]string, map[string]string, error) {
+	plog.Printf("Connecting to AWS %v...", cloud.Name)
+	api, err := aws.New(&aws.Options{
+		Profile: cloud.Profile,
+		Region:  cloud.BucketRegion,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("creating client for %v: %v", cloud.Name, err)
+	}
+
+	f, err := os.Open(imagePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Could not open image file %v: %v", imagePath, err)
+	}
+	defer f.Close()
+
+	s3ObjectPath := fmt.Sprintf("%s/%s/%s", specBoard, specVersion, strings.TrimSuffix(spec.AWS.Image, filepath.Ext(spec.AWS.Image)))
+	s3ObjectURL := fmt.Sprintf("s3://%s/%s", cloud.Bucket, s3ObjectPath)
+
+	snapshot, err := api.FindSnapshot(imageName)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to check for snapshot: %v", err)
+	}
+
+	if snapshot == nil {
+		plog.Printf("Creating S3 object %v...", s3ObjectURL)
+		err = api.UploadObject(f, cloud.Bucket, s3ObjectPath, false)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Error uploading: %v", err)
+		}
+
+		plog.Printf("Creating EBS snapshot...")
+		snapshot, err = api.CreateSnapshot(imageName, s3ObjectURL, aws.EC2ImageFormatVmdk)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to create snapshot: %v", err)
+		}
+	}
+
+	// delete unconditionally to avoid leaks after a restart
+	plog.Printf("Deleting S3 object %v...", s3ObjectURL)
+	err = api.DeleteObject(cloud.Bucket, s3ObjectPath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error deleting S3 object: %v", err)
+	}
+
+	plog.Printf("Creating AMIs from %v...", snapshot.SnapshotID)
+	hvmImageID, err := api.CreateHVMImage(snapshot.SnapshotID, imageName+"-hvm", imageDescription+" (HVM)")
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create HVM image: %v", err)
+	}
+
+	pvImageID, err := api.CreatePVImage(snapshot.SnapshotID, imageName, imageDescription+" (PV)")
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create PV image: %v", err)
+	}
+
+	err = api.CreateTags([]string{hvmImageID, pvImageID}, map[string]string{
+		"Channel": specChannel,
+		"Version": specVersion,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't tag images: %v", err)
+	}
+
+	postprocess := func(imageID string) (map[string]string, error) {
+		if len(cloud.LaunchPermissions) > 0 {
+			if err := api.GrantLaunchPermission(imageID, cloud.LaunchPermissions); err != nil {
+				return nil, err
+			}
+		}
+
+		destRegions := make([]string, 0, len(cloud.Regions))
+		foundBucketRegion := false
+		for _, region := range cloud.Regions {
+			if region != cloud.BucketRegion {
+				destRegions = append(destRegions, region)
+			} else {
+				foundBucketRegion = true
+			}
+		}
+		if !foundBucketRegion {
+			plog.Noticef("AMI will be created in %v but \"plume release\" will not make it public", cloud.BucketRegion)
```

Comment:
This could probably be fatal, right. afaik this shouldn't ever happen.

Also, if we allow this, then the ami map below potentially has a non-public image entry.